### PR TITLE
session: add encoded plan cache to avoid repeated encode plan when recording slow-log and statement (#21139)

### PR DESCRIPTION
**Pull request by ti-srebot**
cherry-pick #21139 to release-4.0

---

Signed-off-by: crazycs520 <crazycs520@gmail.com>

### What problem does this PR solve?

When the SQL `plan` is huge, `EncodePlan ` will also take more time.

This PR adds a cache to avoid repeated encode plan when recording slow-log and statement.

### Related changes

- Need to cherry-pick to the release branch

### Check List 

Tests 

- N/A

Side effects

- No

### Release note 

- N/A

### CODE REVIEW

File: infoschema/tables_test.go [hunk 1]
```diff
@@ -1344,4 +1354,26 @@ func (s *testTableSuite) TestFormatVersion(c *C) {
 		version := infoschema.FormatVersion(v, false)
 		c.Assert(version, Equals, res[i])
 	}
+<<<<<<< HEAD
+=======
```
Comment (djshow832) [inline infoschema/tables_test.go:1358]: Please resolve conflicts. @crazycs520

**Review by AilinKid — COMMENTED** [review]
PTAL

**Review by djshow832 — APPROVED** [review]
LGTM

**Review by AilinKid — APPROVED** [review]
LGTM

**Review by SunRunAway — APPROVED** [review]

**Pull request merged**